controller.go

```diff
@@ -0,0 +1,235 @@
+
+	// Initialize kvm-operator provider.
+	cloud, err := cloudprovider.InitCloudProvider("kvm-operator", "")
+	if err != nil {
+		glog.Fatalf("failed to initialize cloud provider: %v", err)
+	}
+
+	// Create controller instance.
+	controller := newController(kubeClient, cloud, nodeMonitorPeriod)
+
+	// Start the kvm-operator node controller.
+	stop := make(chan struct{})
+	defer close(stop)
+	go controller.Run(stop)
+
+	// Wait forever.
+	select {}
+}
+
+// Creates new instance of controller.
+func newController(
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration) *controller {
+	return &controller{
+		kubeClient:        kubeClient,
+		cloud:             cloud,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+	}
+}
+
+// Run controller function in a loop.
+func (c *controller) Run(stopCh chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Let the workers stop when we are done
+	glog.Info("Starting kvm-operator node controller")
+
+	// Start a loop to periodically check if any nodes have been deleted from cloudprovider
+	go wait.Until(c.MonitorNode, c.nodeMonitorPeriod, stopCh)
+
+	<-stopCh
+	glog.Info("Stopping kvm-operator node controller")
+}
+
+// Monitor node queries the cloudprovider for non-ready nodes and deletes them
+// if they cannot be found in the cloud provider.
+func (c *controller) MonitorNode() {
+	// Get nodes from cloud provider.
+	instances, ok := c.cloud.Instances()
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
+		return
+	}
+
+	// Get nodes known by kubernetes cluster.
+	nodes, err := c.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
+	if err != nil {
+		glog.Errorf("Error monitoring node status: %v", err)
+		return
+	}
+
+	for i := range nodes.Items {
+		var currentReadyCondition *v1.NodeCondition
+		node := &nodes.Items[i]
+
+		glog.Infof("Checking node %s", node.Name)
+
+		// Try to get the current node status
+		// If node status is empty, then kubelet has not posted ready status yet. In this case, process next node.
+		for rep := 0; rep < nodeStatusUpdateRetry; rep++ {
+			_, currentReadyCondition = nodeutilv1.GetNodeCondition(&node.Status, v1.NodeReady)
+			if currentReadyCondition != nil {
+				break
+			}
+			name := node.Name
+			node, err = c.kubeClient.CoreV1().Nodes().Get(name, metav1.GetOptions{})
+			if err != nil {
+				glog.Errorf("Failed while getting a Node to retry updating NodeStatus. Probably Node %s was deleted.", name)
+				break
+			}
+			time.Sleep(retrySleepTime)
+		}
+		if currentReadyCondition == nil {
+			glog.Errorf("Update status of Node %v from Controller exceeds retry count or the Node was deleted.", node.Name)
+			continue
+		}
+		// If the known node status says that Node is NotReady, then check if the node has been removed
+		// from the cloud provider. If node cannot be found in cloudprovider, then delete the node immediately.
+		if currentReadyCondition != nil {
```

Comment:
prev. one `if currentReadyCondition == nil` means we did not get any node status, so continue with NEXT one.

But if there is status `if currentReadyCondition != nil` check further what to do with the node.

For me looks ok. Or i did not get smth

Reply:
After the code below you don't have to check if `currentReadyCondition != nil` because it is.
 
```
if currentReadyCondition == nil {
       continue
}
```

Reply:
I think all is right and it works as expected.

`continue` will just stop current iteration and start new one.

Reply:
Yes, you just don't need this guard `if currentReadyCondition != nil {` and another indentation level.